python/tests/ibd.py

```diff
@@ -0,0 +1,485 @@
+
+    def add_to_chain(self, other):
+        """
+        Adds a SegmentList to the end of the chain currently bookended
+        by SegL_head and SegL_tail.
+        """
+        assert isinstance(other, SegmentList)
+
+        if len(self.SegL_head) == 0:
+            # No segment lists in the chain.
+            self.SegL_head = other
+            self.SegL_tail = other
+        else:
+            # Multiple segment lists in the chain.
+            self.SegL_tail.next = other
+            self.SegL_tail = other
+
+        return True
+
+    def find_ibd_segments(self, min_length=0, max_time=None):
+        """
+        The wrapper for the procedure that calculates IBD segments.
+        """
+
+        # Set up an iterator over the edges in the tree sequence.
+        mygen = iter(self.ts.edges())
+        e = next(mygen)
+
+        # Iterate over the edges.
+        while e is not None:
+            # Process all edges with the same parent node.
+            S_head = None
+            S_tail = None
+            self.current_parent = e.parent
+            self.current_time = self.tables.nodes.time[self.current_parent]
+            if self.max_time is not None and self.current_time > self.max_time:
+                # Stop looking for IBD segments once the
+                # processed nodes are older than the max time.
+                break
+
+            # Create a list of segments, S, bookended by S_head and S_tail.
+            # The list holds all segments that are immediate descendants of
+            # the current parent node being processed.
+            while e is not None and self.current_parent == e.parent:
+                seg = Segment(e.left, e.right, e.child)
+                if S_head is None:
+                    S_head = seg
+                    S_tail = seg
+                else:
+                    S_tail.next = seg
+                    S_tail = S_tail.next
+                if e.id < self.ts.num_edges - 1:
+                    e = next(mygen)
+                    continue
+                else:
+                    e = None
+
+            # Create a list of SegmentList objects, SegL, bookended by SegL_head and
+            # SegL_tail. Each SegmentList corresponds to a segment s in S, and
+            # contains all segments of samples that descend from s.
+            while S_head is not None:
+                u = S_head.node
+                list_to_add = SegmentList()
+                if u in self.samples:
+                    list_to_add + Segment(S_head.left, S_head.right, u)
+                else:
+                    s = self.A[u].head
+                    while s is not None:
+                        intvl = (max(S_head.left, s.left), min(S_head.right, s.right))
+                        if intvl[1] - intvl[0] > 0:
+                            list_to_add + Segment(intvl[0], intvl[1], s.node)
+                        s = s.next
+                self.add_to_chain(list_to_add)
+                S_head = S_head.next
```

Comment:
Here I think in the first loop you iterate over edges, creating `S`. Then in the second loop you process these one at a time, and `S` isn't used again. That means (unless I'm' missing something) that you don't need two loops, even `S` at all?